AutoTools: Improve user flag handling

jjellio opened the pull request:
AutoTool packages are inconconsistent in handling user flags. Some use `set_build_environment` to declare `VAR=VAL` arguments, or declare those variables in `configure_args`. When doing this, the package usually does not append `spec.compiler[cflags]` or other relevant spack flags.

Spack provides some tooling to handling variables set in `setup_build_environment`, but because of the flexibility in Autotools, this does not result in consistent behavior

E.g., See Issue #14430. And Related #19052

This patch provides some sanity checking to configure arguments and enforces that user flags get added.

This patch does not (yet) attempt to remove duplicate flags.

This patch lays the ground work for supporting a CMake-like `build_type` parameter - and having those args consistently applied.

For now, this patch has the variant for `build_type` declared, but the flag handling is not added (Given the code, adding these flags would be simple)

jjellio committed:
AutoTools: Improve user flag handling

AutoTool packages are inconconsistent in handling user flags. Some use `set_build_environment` to declare `VAR=VAL` arguments, or declare those variables in `configure_args`. When doing this, the package usuaully does not append `spec.compiler[cflags]` or other relevant spack flags.

Spack provides some tooling to handling variables set in `setup_build_environment`, but because of the flexibility in Autotools, this does not result in consistent behavior

E.g., See issue #14430. And Related #19052

This patch provides some santity checking to configure arguments and enforces that user flags get added.

This patch does not (yet) attempt to remove duplicate flags.

This patch lays the ground work for supporting a CMake-like `build_type` parameter - and having those args consistently applied.

For now, this patch has the variant for `build_type` decalred, but the flag handling is not added (Given the code, adding these flags would be simple)

jjellio commented:
@alalazo @becker33 - Thoughts on this? There are several packages that do not respect user flags.

I'd really like to get this to the point of supporting `build_type` like CMake. As-is, I have to hack some of the packages I use to get consistent flags applied.

jjellio committed:
Autotools: Add build_type option and fix flag handling

This patch continues the prior in this PR.

Corrected some edge cases in setting flags
Tuned debug messaging
Added build_type functions (final effect is commented out)
Tested with various flag combinations (set both by package, by build_type, and by user)
Handled the cases where the package may not set any FLAG variable, yet the user flags are meaningful.
Improved documentation in code to highlight various design decision

jjellio commented:
Added the `build_type` logic and handled more corner cases. Added better comments and documentation.

As-is this works well for me. I do have some reservations about `build_type` flags. That is, are `-O3` and `-g` portable? I think in most cases they are - but they should be checked for supported toolchains in spack. (A windows toolchain could give hell).  A good fix would be to have in the Autotools source directory, a build_types.yaml, with each supported compilers's defaults for Release, Debug, etc... (min size worries me - I never use it, and I doubt it is portable)

jjellio commented:
For what it's worth, this patch (plus some of the other unaccepted tweaks I've submitted) successfully used Clingo+Spack to build a rather complicated Trilinos package I've been building:

```
atdm-trilinos@develop%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~complex~cuda+empire~ipo+ninja~pic~rocm~shared~sparc~tpls_shared amdgpu_target=none build_type=RelWithDebInfo ci_hostname=redwood cuda_arch=none exec_space=serial host_lapack=libsci arch=cray-rhel8-zen
    ^boost@1.75.0%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" +atomic+chrono~clanglibcpp~container~context~coroutine+date_time~debug+exception~fiber+filesystem+graph+icu+iostreams+locale+log+math~mpi+multithreaded~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout+wave cxxstd=11 patches=2ab6c72d03dec6a4ae20220a9dfd5c8c572c5294252155b85c6874d97c323199,57a8401dee8f52b0342e0c8147a5b2db834e8d8f3fbcbbc5950016bd3e9e1ef0,f2409bfa0e69e44817a5f8799e25c2e9e5ee50876a5aaacefd32fa647b80472f visibility=hidden arch=cray-rhel8-zen
        ^bzip2@1.0.6%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" +shared arch=cray-rhel8-zen
        ^icu4c@67.1%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined cxxstd=11 arch=cray-rhel8-zen
            ^python@3.7.5%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" +bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib build_type=Undefined patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87,ebdca648c9c1d25f586d7e2a495b62e6d91973b55264a13d89eda1beff72ef56 arch=cray-rhel8-zen
        ^zlib@1.2.11%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" +optimize+pic+shared patches=e93f7400712c2814905815204dadbdebcb91dc77dd586f60cbc82efa102fb539 arch=cray-rhel8-zen
    ^cgns@4.1.1%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~base_scope~fortran+hdf5~int64~ipo~legacy~mem_debug+mpi+ninja+parallel+scoping+shared+static~testing build_type=Release arch=cray-rhel8-zen
        ^cmake@3.19.7%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~doc+ncurses~ninja+openssl+ownlibs~qt build_type=Release patches=b48396c0e4f61756248156b6cebe9bc0d7a22228639b47b5aa77c9330588ce88 arch=cray-rhel8-zen
        ^hdf5@1.10.7%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~cxx~debug+fortran+hl~java+mpi+pic+shared+szip~threadsafe api=none build_type=Undefined arch=cray-rhel8-zen
            ^libszip@2.1.1%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined arch=cray-rhel8-zen
            ^mvapich2@2.3.5%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~alloca~cuda~debug+regcache~wrapperrpath build_type=Undefined ch3_rank_bits=32 fabrics=mrail file_systems=auto process_managers=slurm threads=multiple arch=cray-rhel8-zen
            ^numactl@2.0.14%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined patches=4e1d78cbbb85de625bad28705e748856033eaafab92a66dffd383a3d7e00cc94,62fc8a8bf7665a60e8f4c93ebbd535647cebf74198f7afafec4c085a8825c006 arch=cray-rhel8-zen
                ^autoconf@2.69%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined arch=cray-rhel8-zen
                ^automake@1.16.1%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined arch=cray-rhel8-zen
                ^libtool@2.4.6%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined arch=cray-rhel8-zen
                ^m4@1.4.18%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" +sigsegv build_type=Undefined patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=cray-rhel8-zen
        ^ninja@kitware%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  arch=cray-rhel8-zen
    ^cray-libsci@20.03.2.1%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~mpi~openmp+shared arch=cray-rhel8-zen
    ^metis@5.1.0%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~gdb~int64+real64+shared build_type=Release patches=4991da938c1d3a1d3dea78e49bbebecba00273f98df2a656e38b83d55b281da1 arch=cray-rhel8-zen
    ^netcdf-c@4.7.4%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" +dap~hdf4~jna+mpi+parallel-netcdf+pic+shared build_type=Undefined patches=2c88dfbd6d339a0336a43b14a65a1d1df995b853b645e4af612617612a642a53 arch=cray-rhel8-zen
        ^curl@7.76.0%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~darwinssl~gssapi~libssh~libssh2~nghttp2 build_type=Undefined arch=cray-rhel8-zen
            ^libidn2@2.3.0%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined arch=cray-rhel8-zen
                ^libunistring@0.9.10%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined arch=cray-rhel8-zen
                    ^libiconv@1.16%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections"  build_type=Undefined arch=cray-rhel8-zen
            ^openssl@1.1.1c%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~docs+systemcerts arch=cray-rhel8-zen
        ^parallel-netcdf@1.12.1%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~burstbuffer~cxx+fortran+pic+shared build_type=Undefined arch=cray-rhel8-zen
            ^perl@5.26.4%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" +cpanm+shared+threads arch=cray-rhel8-zen
    ^parmetis@4.0.3%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~gdb~int64~ipo+ninja+shared build_type=Release patches=4f892531eb0a807eb1b82e683a416d3e35154a455274cf9b162fb02054d11a5b,50ed2081bc939269689789942067c58b3e522c269269a430d5d34c00edbc5870,704b84f7c7444d4372cb59cca6e1209df4ef3b033bc4ee3cf50f369bce972a9d,7be1563dcde9b2d19e8a51d7908a2db6babe425124fc9f2d25df6b80c91df97f arch=cray-rhel8-zen
    ^superlu-dist@6.4.0%cce@11.0.4 cflags="-g -fPIC -ffunction-sections -fdata-sections" cxxflags="-g -fPIC -ffunction-sections -fdata-sections" fflags="-g -fPIC" ldflags="-Wl,--gc-sections" ~cuda~int64~ipo~openmp+shared build_type=Release cuda_arch=none arch=cray-rhel8-zen
```

In this case I used the default `build_type`, Then tried again after setting my top-level `~/.spack/packages.yaml` with `all` using `variants: build_type=Release` (and it successfully applied 'release' flags to all subsequent AutoTool packages).

After this patch is polished I'll look at CMake packages - they are not being consistent with flags either. But CMake should be pretty easy to deal with since it has a well defined interface to setting flags and such.

haampie commented:
I haven't reviewed this, but just wanted to comment this looks promising :)

jjellio commented:
I need to push a small fix (not adding a flag `VAR=` if spack does not have the flag set). 

For sake of clarity, here is some example output (I've now used this build python3, cmake, trilinos, and their hordes of dependencies )

One test is with spack compiler.yaml setting C/CXX flags to one thing and Fortran flags to another, while leaving ldflags empty
```
cflags: -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections
cxxflags: -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections
fflags: cflags: -mtune=znver1 -march=znver1
# ldflags:
```

For this dependency (cut from a python build) - the package has set no configure arguments with `FOO=BAR`, where `FOO \in {CFLAGS, CXXFLAGS, F<any of them>FLAGS, LDFLAGS }`

The build has a top level `build_type=Release`

```
==> [2021-04-08-05:53:19.073376] Staging archive: /tmp/coe0201/spack-stage/spack-stage-libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib/libbsd-0.10.0.tar.xz
==> [2021-04-08-05:53:19.074962] '/usr/bin/tar' '-oxf' '/tmp/coe0201/spack-stage/spack-stage-libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib/libbsd-0.10.0.tar.xz'
==> [2021-04-08-05:53:19.122217] Created stage in /tmp/coe0201/spack-stage/spack-stage-libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib
==> [2021-04-08-05:53:19.123140] Ran patch() for libbsd
==> [2021-04-08-05:53:19.123316] libbsd: Building libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib [AutotoolsPackage]
==> [2021-04-08-05:53:19.141538] libbsd: Executing phase: 'autoreconf'
==> [2021-04-08-05:53:19.143589] libbsd: Executing phase: 'configure'
==> [2021-04-08-05:53:19.144168] enforce_user_flags:  Adding CFLAGS=-mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections to configure args
==> [2021-04-08-05:53:19.144234] enforce_user_flags:  Adding CXXFLAGS=-mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections to configure args
==> [2021-04-08-05:53:19.144279] enforce_user_flags:  Adding F77FLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-05:53:19.144321] enforce_user_flags:  Adding F90FLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-05:53:19.144360] enforce_user_flags:  Adding FCFLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-05:53:19.144398] enforce_user_flags:  Adding FFLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-05:53:19.144444] enforce_user_flags:  NOT adding: LDFLAGS= to configure args
==> [2021-04-08-05:53:19.145332] '/tmp/coe0201/spack-stage/spack-stage-libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib/spack-src/configure' '--prefix=/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib' 'CFLAGS=-O3 -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections' 'CXXFLAGS=-O3 -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections' 'F77FLAGS=-O3 -mtune=znver1 -march=znver1' 'F90FLAGS=-O3 -mtune=znver1 -march=znver1' 'FCFLAGS=-O3 -mtune=znver1 -march=znver1' 'FFLAGS=-O3 -mtune=znver1 -march=znver1'
```

The sauce, is
```
# In these cases, it correctly adds my spack compiler.yaml flags:
==> [2021-04-08-05:53:19.144168] enforce_user_flags:  Adding CFLAGS=-mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections to configure args
==> [2021-04-08-05:53:19.144234] enforce_user_flags:  Adding CXXFLAGS=-mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections to configure args
==> [2021-04-08-05:53:19.144279] enforce_user_flags:  Adding F77FLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-05:53:19.144321] enforce_user_flags:  Adding F90FLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-05:53:19.144360] enforce_user_flags:  Adding FCFLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-05:53:19.144398] enforce_user_flags:  Adding FFLAGS=-mtune=znver1 -march=znver1 to configure args

# since no ldflags are set, this variable is not appended to ./configure (this is the fix I need to push)
==> [2021-04-08-05:53:19.144444] enforce_user_flags:  NOT adding: LDFLAGS= to configure args
```

I lack `debug` output for the `build_type` setting phase, but you can see 'release' flags are now prepended to the priors when `./configure` is called
```

==> [2021-04-08-05:53:19.145332] '/tmp/coe0201/spack-stage/spack-stage-libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib/spack-src/configure' \
'--prefix=/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/libbsd-0.10.0-tqgnrflheojmcuj7x236y6rousuymxib' \
'CFLAGS=-O3 -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections' \
'CXXFLAGS=-O3 -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections' \
'F77FLAGS=-O3 -mtune=znver1 -march=znver1' \
'F90FLAGS=-O3 -mtune=znver1 -march=znver1' \
'FCFLAGS=-O3 -mtune=znver1 -march=znver1' \
'FFLAGS=-O3 -mtune=znver1 -march=znver1'
```
# The fun cases (and what motivated this patch effort)
In this case, the package sets `CFLAGS`, but does not pass in spack's flags...

My patch could actually become CI testing if you want to impose that packages test for user flags

```
==> [2021-04-08-06:02:12.807355] python: Building python-3.9.2-ofsf7jcam6baj6koxrqprurpkj7rzaal [AutotoolsPackage]
==> [2021-04-08-06:02:12.843222] python: Executing phase: 'autoreconf'
==> [2021-04-08-06:02:12.845524] python: Executing phase: 'configure'
==> [2021-04-08-06:02:12.859406] Warning: enforce_user_flags:  Package set CFLAGS but did not include user flags (cflags)
==> [2021-04-08-06:02:12.859486] Warning: Spack    Flags: -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections
==> [2021-04-08-06:02:12.859538] Warning: Packgage Flags: -fPIC
==> [2021-04-08-06:02:12.859606] Setting CFLAGS=-mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections -fPIC (from configure_args)
==> [2021-04-08-06:02:12.859666] Setting LDFLAGS=-L/usr/lib64 -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/expat-2.2.10-vgomixwodd3hnltffcpktuuxgzfjfgvu/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/gdbm-1.19-e4raldnkc3dzokhvim2oilbmkkreltgo/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/gettext-0.21-rfly7xc3oim7hj4255fdhxnyi7yx7xc3/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/libffi-3.3-hwz3ms2skjghndpnv76ftegut4n5qgvz/lib64 -L/usr/lib64 -L/usr/lib64 -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/readline-8.1-4eigse5ynacamfcllkf77vfsychouzzp/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/sqlite-3.35.3-5ouyfeoaprkovc336h762qzrmrproexz/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/util-linux-uuid-2.36.2-iaze4564tiyt5kjsgvsoipen5twqmnq4/lib -L/usr/lib64 -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/zlib-1.2.11-2fqrr5qfyxy3r6eycm5dtbo4aix4jw5m/lib (from configure_args)
==> [2021-04-08-06:02:12.859738] enforce_user_flags:  Adding CXXFLAGS=-mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections to configure args
==> [2021-04-08-06:02:12.859782] enforce_user_flags:  Adding F77FLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-06:02:12.859821] enforce_user_flags:  Adding F90FLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-06:02:12.859859] enforce_user_flags:  Adding FCFLAGS=-mtune=znver1 -march=znver1 to configure args
==> [2021-04-08-06:02:12.859898] enforce_user_flags:  Adding FFLAGS=-mtune=znver1 -march=znver1 to configure args
```

The awesome is:
```
==> [2021-04-08-06:02:12.859406] Warning: enforce_user_flags:  Package set CFLAGS but did not include user flags (cflags)
==> [2021-04-08-06:02:12.859486] Warning: Spack    Flags: -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections
==> [2021-04-08-06:02:12.859538] Warning: Packgage Flags: -fPIC
==> [2021-04-08-06:02:12.859606] Setting CFLAGS=-mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections -fPIC (from configure_args)
```

In this case, python is setting `-fPIC`, which then sets CFLAGS to only `-fpic`. My patch will actually report where the CFLAG was set from (`setup_build_environment` or `configure_args`). In 100% of the examples here, the setting is all done via `configure_args`.

One idea would be to try and impose that packages handle respecting user flags, but seems kinda sketchy. An option to do that via the wrappers is equally sketchy.  I prefer this approach, because it makes it very very clear what is getting set. As in, you can look at the configure line and immediately know what configure is seeing - not some hocus pocus where `configure` looks good, but internally it's choking because some flag was added behind the scenes by the wrapper.

Now with this bundle of amazement, you get python's configure line not only with user flags, but also top-level imposed `release` flags
```
'/tmp/coe0201/spack-stage/spack-stage-python-3.9.2-ofsf7jcam6baj6koxrqprurpkj7rzaal/spack-src/configure \
--prefix=/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/python-3.9.2-ofsf7jcam6baj6koxrqprurpkj7rzaal \
CPPFLAGS=-I/usr/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/expat-2.2.10-vgomixwodd3hnltffcpktuuxgzfjfgvu/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/gdbm-1.19-e4raldnkc3dzokhvim2oilbmkkreltgo/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/gettext-0.21-rfly7xc3oim7hj4255fdhxnyi7yx7xc3/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/libffi-3.3-hwz3ms2skjghndpnv76ftegut4n5qgvz/include -I/usr/include -I/usr/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/readline-8.1-4eigse5ynacamfcllkf77vfsychouzzp/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/sqlite-3.35.3-5ouyfeoaprkovc336h762qzrmrproexz/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/util-linux-uuid-2.36.2-iaze4564tiyt5kjsgvsoipen5twqmnq4/include -I/usr/include -I/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/zlib-1.2.11-2fqrr5qfyxy3r6eycm5dtbo4aix4jw5m/include \
--enable-optimizations \
--without-pydebug \
--enable-shared \
--without-ensurepip \
--with-openssl=/usr \
--with-dbmliborder=gdbm:bdb:ndbm \
--with-system-expat \
--with-system-ffi \
--enable-loadable-sqlite-extensions \
CFLAGS=-O3 -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sections -fPIC \
LDFLAGS=-L/usr/lib64 -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/expat-2.2.10-vgomixwodd3hnltffcpktuuxgzfjfgvu/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/gdbm-1.19-e4raldnkc3dzokhvim2oilbmkkreltgo/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/gettext-0.21-rfly7xc3oim7hj4255fdhxnyi7yx7xc3/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/libffi-3.3-hwz3ms2skjghndpnv76ftegut4n5qgvz/lib64 -L/usr/lib64 -L/usr/lib64 -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/readline-8.1-4eigse5ynacamfcllkf77vfsychouzzp/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/sqlite-3.35.3-5ouyfeoaprkovc336h762qzrmrproexz/lib -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/util-linux-uuid-2.36.2-iaze4564tiyt5kjsgvsoipen5twqmnq4/lib -L/usr/lib64 -L/home/groups/coegroup/atdm-dev/atdm-spack/software/cray-rhel8-zen/gcc-9.2.0/zlib-1.2.11-2fqrr5qfyxy3r6eycm5dtbo4aix4jw5m/lib \
CXXFLAGS=-O3 -mtune=znver1 -march=znver1 -fdata-sections -ffunction-sec' \
'F77FLAGS=-O3 -mtune=znver1 -march=znver1' \
'F90FLAGS=-O3 -mtune=znver1 -march=znver1' \
'FCFLAGS=-O3 -mtune=znver1 -march=znver1' \
'FFLAGS=-O3 -mtune=znver1 -march=znver1'
```

mwkrentel commented:
I have some concerns, especially for adding this to every autotools
package.  First, let's review the major issues.

(1) The main thing here is to push the decision of compiler flags
(mostly optimization level and debug) up to the spack level.  Left
alone, autotools will decide something on its own, but not always what
you'd expect.  It's better to set this at the spack level and then
tell autotools the answer.

This much is fairly straightforward.  You can either use `build_type`
like cmake, or else separate `opt_flags` and `debug`.  Either way.

(2) The hard part is how you tell autotools what the decision is.
It's hard because you need to pass the flags in a way that the package
expects and not all packages behave the same.

Some packages want CFLAGS= on the command line.  Some want CFLAGS in
the environment.  It's on a package by package basis, there are quirks
and some packages need special treatment.  I'll see if I can find some
specific examples of trouble for you.

(3) As you've discovered, there is a HUGE difference between not
specifying CFLAGS and setting CFLAGS= (empty).  In the first case,
autotools will make its own decision, in the second, you've erased the
optimization and debug flags.

One of the major technical challenges of this PR is to smooth over
this difference.  I want to try this on some of the harder examples,
but I kinda need you to fix this first before I can adequately test it.

----------

My biggest concern is adding this to all 1,000 autotools packages.
It's bound to break something.  Not all "autotools" packages accept
CLFAGS, etc on the configure line.  Some look for this in the
environment.  Some packages already have workarounds for the issues
above and you may cut across their efforts.

I would feel better if you made it an add-on option and let packages
add this individually.  If you do this, then you don't really need the
'Undefined' build type.

Some other things:

1. If you use `build_type` like cmake, then you should use the same
names like cmake: Debug, Release, RelWithDebInfo and MinSizeRel.

2. If it were me, I'd set the Release and RelWithDebInfo opt levels to
-O2, not -O3.  That's the normal autotools default.  Otherwise you're
going to change the value for 1,000 packages and someone is going to
object.

3. I'm guessing the opt values probably do depend on the compiler
family, but I'm not sure.

Anyway, fix the FLAGS= empty vs not setting FLAGS issue and then I'll
try testing some of the harder packages to see what kind of trouble I can make.

jjellio commented:
This does not push this to all autotools - it defaults to 'build_type=Undefined', which does not do anything w.r.t optimizatation flags. The other issue is a defect - user flags get completely ignored.

@mwkrentel - I'll try to push the empty var thing, that actually didn't give me issues - it popped up because in my tests, (compilers.yaml), I intentionally set: cflags, cxxflags (to one thing), ffflags (to something slightly different), and ldflags (unset). The 'bug' is have 'LDFLAGS=' on the configure line, which was more of an annoyance.

As for the CMake naming - it is intended to be the same (if I botched that let me know) the only deviation is that we *need* to have `build_type=Undefined`, that way we preserve existing behavior (as in - set nothing).

We can review/discuss the general problem(s) though
1. Before configure is called, default configure options are applied by `getattr(self, 'configure_flag_args', [])`, which is defined in `flags_to_build_system_args`, which is populated by `set_compiler_environment_variables` (part of build system)
2. Subsequently, the package can define `configure_args` and put whatever there.
3. A missing component is a reconciliation between steps 1 and 2 (even though I haven't see a discrepancy yet - I suspect if the package did `setup_build_environment` with `CFLAGS=blah`, I may be able to see that ) - I'm not really sure
    * Regardless, I added testing step around these two things (1) and (2) to check for flag variables getting set
    * The process I use simply constructs maps (dicts) for each, then can combine and report what configure will actually do.
4. Since I now have clear sets of expected flags (from the package) e.g., CFLAGS=-fpic, I can now unify these two. I.e., if CFLAGS is set by BUILD ENV and CXXFLAGS is set in `configure_args`, I can now reduce that a single map of flag variables and values
5. The important step is then to scan these variables and ensure that if a user has `cxxflags=blah`, that `CFLAGS` has blah
    * and to do this step without reordering flags. To this end, I do tests for inclusion (without reordering or depuplicating). If the user flag is **not** found, then the policy I defined is: **prepend** that could just as easily be *postpend*, but ultimately we have to choose how to add map spack flags `cxxflags` to `CXXFLAGS` (its a policy decision)

Step 6 is included (but disabled) - it adds the `build_type` variant
Step 6 is included because the ability to add a consistent optimization level is equivalent to resolving the prior steps (you'd have to reason out flags from the ENV or configure line, *and* you'd need to resolve any conflicts)
Step 6 is then a rather simple thing to do after 1-5, because by step 5, you have a single map of `{flag_variable : values}`


As for optimization flag and toolchain/vendors - I mention that in the top post. In the code, I've speculated that you could just define a yaml. (dict) of 
```yaml
build_types:
  gcc:
    Release:
    Debug:
  cce:
... (whatever the proper names are for spack's supported compilers)
```
This would clearly isolate these definitions to say `build_system/build_types` (you could make it module in the python sense) - or perhaps LLNL or SPACK (python packages) already have the tooling. Given a key (which we choose conviently to be spack's compiler toolchain name + a level that maps exactly to what CMake uses, you get something like `module::get_flags('cce','Release')`

As of right now, the locations and definitions of these things are kinda ambiguous because we haven't discussed what to do about them. (but we should)


Does that help? (these issues are discussed in the code as well)

jjellio committed:
Do not set variables that neither SPACK nor the package have set

jjellio commented:
I've pushed the correction - the `build_type` stuff is there, but the call to enable it is commented out.

e.g.,
```diff
@@ -800,7 +807,7 @@ def configure(self, spec, prefix):
         # DISABLED initially - included to demostrate why the prior effort
         # make this clean to implement
         #
         # self._configure_apply_build_type(spec, unified_flags)
```

jjellio commented:
@mwkrentel I think a policy where the flags are enforced in the area they were already set seems fine (to resolve build ENV vs command line). That shouldn't be hard to do since I have that tracked (I'll just need to keep up with the mapping of which var came from where - so far all have come from cmd line).  I can probably add / push that pretty fast (today/tomorrow)

# I'll be adding better env support and reorganizing some code with a commit today

jjellio committed:
include env flags

jjellio committed:
Address numerous issues.

1. `env` variables are corrected parsed (previously I was only parsing 2 of 3 places that flags can come from)
2. I track locations that variables come from - the top-level `configure` is getting a little busy
    *  The top level defines a mapping between sources and what to do - this has to be autotools dependen, though I'd like this functionality to work in other build types
3. control flow is a little different. I work through all 3 possible **sources**: `flags_to_build_system_args` (populates `configure_flag_args`), `setup_build_environment`, and `configure_args`.
    * From each source, I capture flag variables (from a whitelist)
    * For `configure_flag_args` or `configure_args` I capture the non-whitelisted arguments
4. I next **reconcile** the flag information I obtained for each source.
    * this includes checking for duplicates (i.e., setting CFLAGS in both ENV and on the command line)
    * If duplicates I also test if the package simply replicated the flags in both places
    * I then produce a single unified map of flag variables and values, while also retaining where those variables came from
5. With the `unified` map, I then `enforce_user_flags`
    *  This checks if a `VARNAME` was set that maps to spack flag variable (CFLAGS -> cflags)
    * Checks that if VARNAME was set, that spack (user) flags are respected (a motivation of this PR)
    * This detects if the package did not set a given VARNAME, but the user did set the corresponding spack variable (another motivator for this PR)
    * Handles the case that neither a VARNAME or associated spack varaible was set (and does nothing)
    * returns the final set of flags complete with user flags prepended
6. [optional] enable the `build_type` feature - this will adjust the final (user + package) set of flags based on the `build_type` variant - the default is `Undefined`, which means nothing is done.
7. Sets the values
    *  This uses location information and a mapping dict that explains what to do with those locations. (currently two choices: `env` or `return`)
    * env does what you think, it puts the variable in the `build_environment` (and maps to variables that were set via `setup_build_environment`
    * return means the values are placed back into `flag_map` and returned - and will the be postpended to `configure_args`

jjellio commented:
Will be pushing a some changes to address Mark's comments @mwkrentel 
Some changes:
1. `env` variables are corrected parsed (previously I was only parsing 2 of 3 places that flags can come from)
2. I track locations that variables come from - the top-level `configure` is getting a little busy
    *  The top level defines a mapping between sources and what to do - this has to be autotools dependen, though I'd like this functionality to work in other build types
3. control flow is a little different. I work through all 3 possible **sources**: `flags_to_build_system_args` (populates `configure_flag_args`), `setup_build_environment`, and `configure_args`.
    * From each source, I capture flag variables (from a whitelist)
    * For `configure_flag_args` or `configure_args` I capture the non-whitelisted arguments
4. I next **reconcile** the flag information I obtained for each source.
    * this includes checking for duplicates (i.e., setting CFLAGS in both ENV and on the command line)
    * If duplicates I also test if the package simply replicated the flags in both places
    * I then produce a single unified map of flag variables and values, while also retaining where those variables came from
5. With the `unified` map, I then `enforce_user_flags`
    *  This checks if a `VARNAME` was set that maps to spack flag variable (CFLAGS -> cflags)
    * Checks that if VARNAME was set, that spack (user) flags are respected (a motivation of this PR)
    * This detects if the package did not set a given VARNAME, but the user did set the corresponding spack variable (another motivator for this PR)
    * Handles the case that neither a VARNAME or associated spack varaible was set (and does nothing)
    * returns the final set of flags complete with user flags prepended
6. [optional] enable the `build_type` feature - this will adjust the final (user + package) set of flags based on the `build_type` variant - the default is `Undefined`, which means nothing is done.
7. Sets the values
    *  This uses location information and a mapping dict that explains what to do with those locations. (currently two choices: `env` or `return`)
    * env does what you think, it puts the variable in the `build_environment` (and maps to variables that were set via `setup_build_environment`
    * return means the values are placed back into `flag_map` and returned - and will the be postpended to `configure_args`

Its worth nothing, virutally no packages are calling `flags_to_build_system_args` (one of the possible locations to do this stuff)
A single package does (but it's cmake based). It doesn't matter because I handled the case anyway


I've written a dummy test package that can be used to test basic functionality
The variants will set pipe `|` delimited variables.
I also have some variables set in `compilers.yaml`
```
spack -d --verbose  install autotools-test-compiler-flags set_in_build_env='CFLAGS=A|FCFLAGS=B' set_in_configure_args='CXXFLAGS=A|LDFLAGS=B' 
...
==> [2021-04-11-20:13:34.631648] [setup_build_environment] Setting CFLAGS=A
==> [2021-04-11-20:13:34.632631] [setup_build_environment] Setting FCFLAGS=B
==> [2021-04-11-20:13:45.155977] '/opt/cray/pe/craype/default/bin/CC' '-g' '-fPIC' '-ffunction-sections' '-fdata-sections' '-v' '/tmp/spack-implicit-link-infolzr0b1f9/main.c' '-o' '/tmp/spack-implicit-link-infolzr0b1f9/output'
==> [2021-04-11-20:13:45.230530] parsing implicit link info
...
==> [2021-04-11-20:13:45.269026] [configure_args] Setting CXXFLAGS=A
==> [2021-04-11-20:13:45.269090] [configure_args] Setting LDFLAGS=B
==> [2021-04-11-20:13:45.280583] autotools-test-compiler-flags: Executing phase: 'autoreconf'
==> [2021-04-11-20:13:45.282671] autotools-test-compiler-flags: Executing phase: 'configure'
==> [2021-04-11-20:13:45.282809] [gather_flag_args]: Discovering flag variables from configure_flag_args
==> [2021-04-11-20:13:45.283999] [gather_flag_args]: Discovering flag variables from build_environment
==> [2021-04-11-20:13:45.284216] [configure_args] Setting CXXFLAGS=A
==> [2021-04-11-20:13:45.284257] [configure_args] Setting LDFLAGS=B
==> [2021-04-11-20:13:45.284297] [gather_flag_args]: Discovering flag variables from configure_args
==> [2021-04-11-20:13:45.284364] [reconcile_flag_vars]: build_environment: CFLAGS=A
==> [2021-04-11-20:13:45.284404] [reconcile_flag_vars]: build_environment: FCFLAGS=B
==> [2021-04-11-20:13:45.284440] [reconcile_flag_vars]: configure_args: CXXFLAGS=A
==> [2021-04-11-20:13:45.284475] [reconcile_flag_vars]: configure_args: LDFLAGS=B
==> [2021-04-11-20:13:45.284526] [reconcile_flag_vars]: Final set of flag vars from build_environment : {FCFLAGS,CFLAGS}
==> [2021-04-11-20:13:45.284566] [reconcile_flag_vars]: Final set of flag vars from configure_args : {LDFLAGS,CXXFLAGS}
==> [2021-04-11-20:13:45.284603] [reconcile_flag_vars]: Final set of flag vars from configure_flag_args : {}
==> [2021-04-11-20:13:45.284644]   From [build_environment] Adding FCFLAGS=B
==> [2021-04-11-20:13:45.284682]   From [build_environment] Adding CFLAGS=A
==> [2021-04-11-20:13:45.284717]   From [configure_args] Adding LDFLAGS=B
==> [2021-04-11-20:13:45.284752]   From [configure_args] Adding CXXFLAGS=A
==> [2021-04-11-20:13:45.284848] Warning: [enforce_user_flags]: build_environment set FCFLAGS but did not include user flags (fflags)
==> [2021-04-11-20:13:45.284891] Warning: Spack    Flags: -g -fPIC
==> [2021-04-11-20:13:45.284925] Warning: Packgage Flags: B
==> [2021-04-11-20:13:45.284973] Warning: [enforce_user_flags]: build_environment set CFLAGS but did not include user flags (cflags)
==> [2021-04-11-20:13:45.285011] Warning: Spack    Flags: -g -fPIC -ffunction-sections -fdata-sections
==> [2021-04-11-20:13:45.285044] Warning: Packgage Flags: A
==> [2021-04-11-20:13:45.285086] Warning: [enforce_user_flags]: configure_args set CXXFLAGS but did not include user flags (cxxflags)
==> [2021-04-11-20:13:45.285123] Warning: Spack    Flags: -g -fPIC -ffunction-sections -fdata-sections
==> [2021-04-11-20:13:45.285156] Warning: Packgage Flags: A
==> [2021-04-11-20:13:45.285201] [enforce_unset_user_flags]: Adding F77FLAGS=-g -fPIC because user set fflags
==> [2021-04-11-20:13:45.285240] [enforce_unset_user_flags]: Adding F90FLAGS=-g -fPIC because user set fflags
==> [2021-04-11-20:13:45.285277] [enforce_unset_user_flags]: Adding FFLAGS=-g -fPIC because user set fflags
==> [2021-04-11-20:13:45.285331] [set_flags_in_preferred_src]: Setting in environment: FCFLAGS=-g -fPIC B
==> [2021-04-11-20:13:45.285380] [set_flags_in_preferred_src]: Setting in environment: CFLAGS=-g -fPIC -ffunction-sections -fdata-sections A
==> [2021-04-11-20:13:45.285418] [set_flags_in_preferred_src]: adding to configure_args: LDFLAGS=B
==> [2021-04-11-20:13:45.285451] [set_flags_in_preferred_src]: adding to configure_args: CXXFLAGS=-g -fPIC -ffunction-sections -fdata-sections A
==> [2021-04-11-20:13:45.285483] [set_flags_in_preferred_src]: adding to configure_args: F77FLAGS=-g -fPIC
==> [2021-04-11-20:13:45.285514] [set_flags_in_preferred_src]: adding to configure_args: F90FLAGS=-g -fPIC
==> [2021-04-11-20:13:45.285544] [set_flags_in_preferred_src]: adding to configure_args: FFLAGS=-g -fPIC
==> [2021-04-11-20:13:45.286246] '/tmp/coe0201/spack-stage/spack-stage-autotools-test-compiler-flags-1.0-mh5slzwvgibabxksg7krsjfl27qp5txp/spack-src/configure' '--prefix=/home/users/coe0201/src/spack/opt/spack/cray-rhel8-zen/cce-11.0.4/autotools-test-compiler-flags-1.0-mh5slzwvgibabxksg7krsjfl27qp5txp' 'LDFLAGS=B' 'CXXFLAGS=-g -fPIC -ffunction-sections -fdata-sections A' 'F77FLAGS=-g -fPIC' 'F90FLAGS=-g -fPIC' 'FFLAGS=-g -fPIC'
```

In this

jjellio committed:
Dummy package for testing flag variables

E.g.,
autotools-test-compiler-flags set_in_build_env='CFLAGS=A|FCFLAGS=B' set_in_configure_args='CXXFLAGS=A|LDFLAGS=B'

mwkrentel commented:
@jjellio Ok, I'll try it with some packages and see what happens.

jjellio commented:
I'll need to adjust the flag adding code - originally I had a conditional if `LDFLAGS` - then do not merge. I suspect, I'll need to uniformly prepend or postpend the entire user specified flags (as a string). There's just too many pitfalls with detecting whether a single flag from a string of flags can be omitted.

For example:
`cxxflags = '-fp-model precise'` (You don't want that mixed up)

LDFLAGS, you're more likely to have the case the package sets `-Wl,blah something important after blah`, and if `ldflags=-Wl,blah something else` you'd create a huge mess. (this could happen with linker groups - I don't think users would set this stuff normally, but they could if they wanted.

It's really important to preserve the full ordering of that - my original code did it, but the pushed version I let it slip. This is an easy change, and can result in replicated flags.

mwkrentel commented:
@jjellio Well, the first thing is that the bug of setting CFLAGS= (empty) on
the command line with the default `build_type=Undefinded` is back.

I know you want to say that the patch is a no-op with the default
`build_type=Undefinded`, but it's things like this that make me
question whether that's really true.

Anyway, fix this and then I'll resume testing.

jjellio commented:
Mark - setting an **empty** variable wasn't the bug. (I guess I can check that too).

The issue was if the user didn't define the variable, setting that variable (in my code) as empty (if compiler.yaml had `# ldflags`, then it would set `LDFLAGS=` (because I was setting all spack variables whether the user provided a value or not).


You are saying if the user sets an empty variable, should that be propagated (I think currently it will - but I didn't test that).

My intent is enforce the behavior that **if** the users sets a spack flag, **then** the requisite autotools variable should reflect whatever they set it at (even if empty)

FYI - I believe if you have `compilers.yaml` defined with an empty variable it will choke

mwkrentel commented:
What I mean is that if you build this spec:

```
libsigsegv@2.12%gcc@7.3.1 build_type=Undefined arch=linux-fedora26-westmere
```

Then you invoke the libsigsegv configure line as:

```
'/home/krentel/autotools/spack-jjellio/build-stage/spack-stage-libsigsegv-2.12-siwarlty7m7ax3o7p2f2ycqtw3ctiqw6/spack-src/configure' '--prefix=/home/krentel/autotools/install-elliot/linux-fedora26-westmere/gcc-7.3.1/libsigsegv-2.12-siwarlty7m7ax3o7p2f2ycqtw3ctiqw6' '--enable-shared' 'CFLAGS=' 'CXXFLAGS=' 'FFLAGS=' 'FCFLAGS=' 'F90FLAGS=' 'F77FLAGS='
```

As I mentioned above and in #14430, this is very wrong because it
changes CFLAGS from the autotools default `-g -O2` to empty and builds
with NO optimization.

If you want `Undefined` to be the do-no-harm no-op, then you can't
change the configure line at all.

jjellio commented:
Thanks Mark (this will help alot) - I'll give that a try and fix it up when I get back to this. (probably tomorrow)

mwkrentel commented:
It's Mark, not Mike.  As I mentioned above, the safer thing is to make
this an optional subclass and let packages enable it on a package by
package basis.  That way, you're SURE it's a no-op.

Btw, it's never a no-op because you're always adding a new variant.

jjellio commented:
Mark, the CMake-like build_system is really secondary to the patches. The point is to provide consistent flag handling (which spack is not doing).

The following should produce identical behaviors:
```
#!/bin/sh
CFLAGS=A
./configure

spack install foo+fpic cflags=A
```

Currently, the first will configure w/CFLAGS=A
The spack version will in most cases not because as soon as you activate a variant that sets any flags, the spack (top level) flags are getting ignored.

The meat of the patch is enforcing consistent flag behavior - the `build_type` thing is something that depends on the effort (but is not the motivator).

You could consider the `build_type` thing a bug resolution as well, because by default `./configure` will choose some optimization level - but the moment you append `CFLAGS=whatever`, `configure` usually will respect that. E.g., my builds tank from any optimizations to **none** - Some Packages, e.g., HDF5 provide tooling for *production* - but they are the exception not the norm.

This is a real problem imo. Both from a user perspective of libraries not getting built as intended - and from an admin or system evaluation angle (something I care about). It matters to me that HDF5 will build w/`-g -O0`, and it matters it builds with `-g -O3 -mtune=something expensive`  The top-level flags enable resolve most problems, and the `build_type` makes it cleaner.

I'm coming at spack after using a `bash` script that sets common top-level flags (e.g., `-mtune` `-march`, `-g`, `-ffunction-sections`, etc..)  My experience with spack has been those flags when set via `compilers.yaml` are mostly ignored.

jjellio committed:
Fix build_type behavior to only set flags if the type is not Undefined

jjellio commented:
Fix incoming

Unset: cflags
Set: cxxflags
Set: fflags
Unset: ldflags

Results

build_type | configure
--- |---
Not set | `'CXXFLAGS=-Werror' 'F77FLAGS=-Wall' 'F90FLAGS=-Wall' 'FCFLAGS=-Wall' 'FFLAGS=-Wall'`
Undefined | `'CXXFLAGS=-Werror' 'F77FLAGS=-Wall' 'F90FLAGS=-Wall' 'FCFLAGS=-Wall' 'FFLAGS=-Wall'`
Release | `'CXXFLAGS=-O3 -Werror' 'F77FLAGS=-O3 -Wall' 'F90FLAGS=-O3 -Wall' 'FCFLAGS=-O3 -Wall' 'FFLAGS=-O3 -Wall' 'CFLAGS=-O3'`

This seems like expected behavior to me. It will set all of the Fortran FLAG variables, even if you set only one - because that's how spack currently works (there is no distinction)

If you enable 'Release' it will set every variable - because it doesn't know which variables the package actually supports. The only way to resolve that would be to run `./configure --help` hope it prints the supported variables, then use those to create the map of variables to set (that may not be hard to do, but I am not sure if how AutoTools figures out what to print down there and whether it is consistent) - even if I did that, I'd still have to take the subset of known variables that have a spack mapping (CFLAGS, CXXFLAGS, etc...)

Adding a feature to spack (in autotools) that parsed 'influential' variables from `configure --help` sounds like a good idea, but who knows. It would enhance the 

```yaml
compilers:
- compiler:
    spec: gcc@8.3.1
    modules: []
    operating_system: rhel8
    target: any
    paths:
      cc:  /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc:  /usr/bin/gfortran
    flags:
      cflags:
      cxxflags: -Werror
      fflags:  -Wall
      ldflags:
    environment: {}
    extra_rpaths: []
```
```terminal
 spack -d --verbose install  libsigsegv@2.12%gcc@8.3.1 build_type=Undefined
```

jjellio committed:
remove fixme so this will not bug out the error checks

jjellio committed:
fix logic error: set flags from source when all sources are the same - was not using source mapping!
